File: leetcode_problems/cpp/lc242_valid_anagram.cpp
#include <string>
#include <iostream>
using std::string;

class Solution
{
public:
    bool run(string s, string t)
    {
        if (s.size() != t.size())
        {
            return false;
        }

        const int N = 26;
        int mp1[N] = {};
        int mp2[N] = {};

        for (int index = 0; index < s.size(); index++)
        {
            mp1[s[index] - 'a']++;
            mp2[t[index] - 'a']++;
        }

        for (int index = 0; index < N; index++)
        {
            if (mp1[index] != mp2[index])
            {
                return false;
            }
        }

        return true;
    }
};

int main()
{
    string s = "anagram";
    string t = "nagaram";

    Solution sol;
    bool answer = sol.run(s, t);

    std::cout << answer << std::endl;
}

File: leetcode_problems/python/lc242_valid_anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        m = dict()
        for c in s:
            m[c] = m.get(c, 0) + 1

        for c in t:
            if c not in m or m[c] <= 0:
                return False
            if m[c] == 1:
                del m[c]
                continue
            m[c] -= 1

        return not m


if __name__ == "__main__":
    s = "anagram"
    t = "nagaram"
    print(Solution().isAnagram(s, t))
